// @redocly/openapi-cli/dist/redocly/query.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = query;
exports.RequestError = void 0;

var _nodeFetch = _interopRequireDefault(require("node-fetch"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const GRAPHQL_ENDPOINT = process.env.REDOCLY_DOMAIN ? `https://api.${process.env.REDOCLY_DOMAIN}/graphql` : 'https://api.redoc.ly/graphql';

async function query(queryString, variables = {}, headers = {}, debugInfo = '') {
  // eslint-disable-next-line no-param-reassign
  headers = { ...headers,
    'Content-Type': 'application/json'
  };
  const gQLResponse = await (0, _nodeFetch.default)(GRAPHQL_ENDPOINT, {
    method: 'POST',
    headers,
    body: JSON.stringify({
      query: queryString,
      variables
    })
  });

  if (!gQLResponse.ok) {
    throw new RequestError(`Failed to execute query: ${gQLResponse.status}`, 500, debugInfo);
  }

  const response = await gQLResponse.json();

  if (response.errors && response.errors.length) {
    throw new RequestError(`Query failed: ${response.errors[0].message}`, 500, debugInfo);
  }

  return response.data;
}

class RequestError extends Error {
  constructor(message, statusCode = 500, debugInfo = '') {
    super(message);
    this.statusCode = statusCode;
    this.debugInfo = debugInfo;
  }

}

exports.RequestError = RequestError;
