// @redocly/openapi-cli/dist/types/OAS2/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "OAS2Root", {
  enumerable: true,
  get: function () {
    return _OAS2Root.default;
  }
});
Object.defineProperty(exports, "OAS2Info", {
  enumerable: true,
  get: function () {
    return _OAS2Info.default;
  }
});
Object.defineProperty(exports, "OAS2License", {
  enumerable: true,
  get: function () {
    return _OAS2License.default;
  }
});
Object.defineProperty(exports, "OAS2Contact", {
  enumerable: true,
  get: function () {
    return _OAS2Contact.default;
  }
});
Object.defineProperty(exports, "OAS2Definitions", {
  enumerable: true,
  get: function () {
    return _OAS2Definitions.default;
  }
});
Object.defineProperty(exports, "OAS2Schema", {
  enumerable: true,
  get: function () {
    return _OAS2Schema.default;
  }
});
Object.defineProperty(exports, "OAS2SchemaMap", {
  enumerable: true,
  get: function () {
    return _OAS2SchemaMap.default;
  }
});
Object.defineProperty(exports, "OAS2ExternalDocumentation", {
  enumerable: true,
  get: function () {
    return _OAS2ExternalDocumentation.default;
  }
});
Object.defineProperty(exports, "OAS2Tag", {
  enumerable: true,
  get: function () {
    return _OAS2Tag.default;
  }
});
Object.defineProperty(exports, "OAS2XML", {
  enumerable: true,
  get: function () {
    return _OAS2XML.default;
  }
});
Object.defineProperty(exports, "OAS2Parameter", {
  enumerable: true,
  get: function () {
    return _OAS2Parameter.default;
  }
});
Object.defineProperty(exports, "OAS2Items", {
  enumerable: true,
  get: function () {
    return _OAS2Items.default;
  }
});
Object.defineProperty(exports, "OAS2Header", {
  enumerable: true,
  get: function () {
    return _OAS2Header.default;
  }
});
Object.defineProperty(exports, "OAS2Headers", {
  enumerable: true,
  get: function () {
    return _OAS2Headers.default;
  }
});
Object.defineProperty(exports, "OAS2Response", {
  enumerable: true,
  get: function () {
    return _OAS2Response.default;
  }
});
Object.defineProperty(exports, "OAS2Example", {
  enumerable: true,
  get: function () {
    return _OAS2Example.default;
  }
});
Object.defineProperty(exports, "OAS2Responses", {
  enumerable: true,
  get: function () {
    return _OAS2Responses.default;
  }
});
Object.defineProperty(exports, "OAS2Scopes", {
  enumerable: true,
  get: function () {
    return _OAS2Scopes.default;
  }
});
Object.defineProperty(exports, "OAS2SecurityRequirement", {
  enumerable: true,
  get: function () {
    return _OAS2SecurityRequirement.default;
  }
});
Object.defineProperty(exports, "OAS2SecurityScheme", {
  enumerable: true,
  get: function () {
    return _OAS2SecurityScheme.default;
  }
});
Object.defineProperty(exports, "OAS2SecurityDefinitions", {
  enumerable: true,
  get: function () {
    return _OAS2SecurityDefinitions.default;
  }
});
Object.defineProperty(exports, "OAS2Parameters", {
  enumerable: true,
  get: function () {
    return _OAS2Parameters.default;
  }
});
Object.defineProperty(exports, "OAS2PathItem", {
  enumerable: true,
  get: function () {
    return _OAS2PathItem.default;
  }
});
Object.defineProperty(exports, "OAS2Operation", {
  enumerable: true,
  get: function () {
    return _OAS2Operation.default;
  }
});
Object.defineProperty(exports, "OAS2Paths", {
  enumerable: true,
  get: function () {
    return _OAS2Paths.default;
  }
});

var _OAS2Root = _interopRequireDefault(require("./OAS2Root"));

var _OAS2Info = _interopRequireDefault(require("./OAS2Info"));

var _OAS2License = _interopRequireDefault(require("./OAS2License"));

var _OAS2Contact = _interopRequireDefault(require("./OAS2Contact"));

var _OAS2Definitions = _interopRequireDefault(require("./OAS2Definitions"));

var _OAS2Schema = _interopRequireDefault(require("./OAS2Schema"));

var _OAS2SchemaMap = _interopRequireDefault(require("./OAS2SchemaMap"));

var _OAS2ExternalDocumentation = _interopRequireDefault(require("./OAS2ExternalDocumentation"));

var _OAS2Tag = _interopRequireDefault(require("./OAS2Tag"));

var _OAS2XML = _interopRequireDefault(require("./OAS2XML"));

var _OAS2Parameter = _interopRequireDefault(require("./OAS2Parameter"));

var _OAS2Items = _interopRequireDefault(require("./OAS2Items"));

var _OAS2Header = _interopRequireDefault(require("./OAS2Header"));

var _OAS2Headers = _interopRequireDefault(require("./OAS2Headers"));

var _OAS2Response = _interopRequireDefault(require("./OAS2Response"));

var _OAS2Example = _interopRequireDefault(require("./OAS2Example"));

var _OAS2Responses = _interopRequireDefault(require("./OAS2Responses"));

var _OAS2Scopes = _interopRequireDefault(require("./OAS2Scopes"));

var _OAS2SecurityRequirement = _interopRequireDefault(require("./OAS2SecurityRequirement"));

var _OAS2SecurityScheme = _interopRequireDefault(require("./OAS2SecurityScheme"));

var _OAS2SecurityDefinitions = _interopRequireDefault(require("./OAS2SecurityDefinitions"));

var _OAS2Parameters = _interopRequireDefault(require("./OAS2Parameters"));

var _OAS2PathItem = _interopRequireDefault(require("./OAS2PathItem"));

var _OAS2Operation = _interopRequireDefault(require("./OAS2Operation"));

var _OAS2Paths = _interopRequireDefault(require("./OAS2Paths"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
