// @redocly/openapi-cli/dist/types/OAS2/OAS2Root.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.OAS2Root = void 0;

var _OAS2Info = require("./OAS2Info");

var _OAS2ExternalDocumentation = require("./OAS2ExternalDocumentation");

var _OAS2Tag = require("./OAS2Tag");

var _OAS2SecurityDefinitions = require("./OAS2SecurityDefinitions");

var _OAS2Definitions = require("./OAS2Definitions");

var _OAS2Parameters = require("./OAS2Parameters");

var _OAS2Responses = require("./OAS2Responses");

var _OAS2SecurityRequirement = require("./OAS2SecurityRequirement");

var _OAS2Paths = require("./OAS2Paths");

const OAS2Root = {
  name: 'OAS2Root',
  isIdempotent: true,
  properties: {
    info: _OAS2Info.OAS2Info,
    paths: _OAS2Paths.OAS2Paths,
    definitions: _OAS2Definitions.OAS2Definitions,
    parameters: _OAS2Parameters.OAS2Parameters,
    responses: _OAS2Responses.OAS2Responses,
    securityDefinitions: _OAS2SecurityDefinitions.OAS2SecurityDefinitions,
    security: _OAS2SecurityRequirement.OAS2SecurityRequirement,
    externalDocs: _OAS2ExternalDocumentation.OAS2ExternalDocumentation,
    tags: _OAS2Tag.OAS2Tag,
    swagger: null,
    host: null,
    basePath: null,
    schemes: null,
    consumes: null,
    produces: null
  }
};
exports.OAS2Root = OAS2Root;
var _default = OAS2Root;
exports.default = _default;
