// @redocly/openapi-cli/dist/types/OAS2/OAS2Schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.OAS2Schema = void 0;

var _OAS2SchemaMap = require("./OAS2SchemaMap");

var _OAS2ExternalDocumentation = require("./OAS2ExternalDocumentation");

var _OAS2XML = require("./OAS2XML");

const OAS2Schema = {
  name: 'OAS2Schema',
  isIdempotent: false,
  properties: {
    allOf() {
      return OAS2Schema;
    },

    items() {
      return OAS2Schema;
    },

    // additional oas2 fields
    additionalProperties() {
      return OAS2Schema;
    },

    xml: _OAS2XML.OAS2XML,
    externalDocs: _OAS2ExternalDocumentation.OAS2ExternalDocumentation,
    properties: _OAS2SchemaMap.OAS2SchemaMap,
    discriminator: null,
    readOnly: null,
    example: null,
    // default JSON schema fields
    title: null,
    description: null,
    default: null,
    maximum: null,
    exclusiveMaximum: null,
    minimum: null,
    exclusiveMinimum: null,
    maxLength: null,
    minLength: null,
    pattern: null,
    multipleOf: null,
    minItems: null,
    maxItems: null,
    uniqueItems: null,
    maxProperties: null,
    minProperties: null,
    required: null,
    enum: null,
    type: null,
    format: null
  },
  resolvableScalars: ['description']
};
exports.OAS2Schema = OAS2Schema;
var _default = OAS2Schema;
exports.default = _default;
