// @redocly/openapi-cli/dist/types/OAS3/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AuthorizationCodeOpenAPIFlow", {
  enumerable: true,
  get: function () {
    return _AuthorizationCodeOpenAPIFlow.default;
  }
});
Object.defineProperty(exports, "ClientCredentialsOpenAPIFlow", {
  enumerable: true,
  get: function () {
    return _ClientCredentialsOpenAPIFlow.default;
  }
});
Object.defineProperty(exports, "ImplicitOpenAPIFlow", {
  enumerable: true,
  get: function () {
    return _ImplicitOpenAPIFlow.default;
  }
});
Object.defineProperty(exports, "OpenAPICallback", {
  enumerable: true,
  get: function () {
    return _OpenAPICallback.default;
  }
});
Object.defineProperty(exports, "OpenAPICallbackMap", {
  enumerable: true,
  get: function () {
    return _OpenAPICallbackMap.default;
  }
});
Object.defineProperty(exports, "OpenAPIComponents", {
  enumerable: true,
  get: function () {
    return _OpenAPIComponents.default;
  }
});
Object.defineProperty(exports, "OpenAPIContact", {
  enumerable: true,
  get: function () {
    return _OpenAPIContact.default;
  }
});
Object.defineProperty(exports, "OpenAPIDiscriminator", {
  enumerable: true,
  get: function () {
    return _OpenAPIDiscriminator.default;
  }
});
Object.defineProperty(exports, "OpenAPIEncoding", {
  enumerable: true,
  get: function () {
    return _OpenAPIEncoding.default;
  }
});
Object.defineProperty(exports, "OpenAPIEncodingsMap", {
  enumerable: true,
  get: function () {
    return _OpenAPIEncodingsMap.default;
  }
});
Object.defineProperty(exports, "OpenAPIExample", {
  enumerable: true,
  get: function () {
    return _OpenAPIExample.default;
  }
});
Object.defineProperty(exports, "OpenAPIExampleMap", {
  enumerable: true,
  get: function () {
    return _OpenAPIExampleMap.default;
  }
});
Object.defineProperty(exports, "OpenAPIExternalDocumentation", {
  enumerable: true,
  get: function () {
    return _OpenAPIExternalDocumentation.default;
  }
});
Object.defineProperty(exports, "OpenAPIFlows", {
  enumerable: true,
  get: function () {
    return _OpenAPIFlows.default;
  }
});
Object.defineProperty(exports, "OpenAPIHeader", {
  enumerable: true,
  get: function () {
    return _OpenAPIHeader.default;
  }
});
Object.defineProperty(exports, "OpenAPIHeaderMap", {
  enumerable: true,
  get: function () {
    return _OpenAPIHeaderMap.default;
  }
});
Object.defineProperty(exports, "OpenAPIInfo", {
  enumerable: true,
  get: function () {
    return _OpenAPIInfo.default;
  }
});
Object.defineProperty(exports, "OpenAPILicense", {
  enumerable: true,
  get: function () {
    return _OpenAPILicense.default;
  }
});
Object.defineProperty(exports, "OpenAPILink", {
  enumerable: true,
  get: function () {
    return _OpenAPILink.default;
  }
});
Object.defineProperty(exports, "OpenAPILinkMap", {
  enumerable: true,
  get: function () {
    return _OpenAPILinkMap.default;
  }
});
Object.defineProperty(exports, "OpenAPIMediaObject", {
  enumerable: true,
  get: function () {
    return _OpenAPIMediaObject.default;
  }
});
Object.defineProperty(exports, "OpenAPIMediaTypeObject", {
  enumerable: true,
  get: function () {
    return _OpenAPIMediaTypeObject.default;
  }
});
Object.defineProperty(exports, "OpenAPIOperation", {
  enumerable: true,
  get: function () {
    return _OpenAPIOperation.default;
  }
});
Object.defineProperty(exports, "OpenAPIParameter", {
  enumerable: true,
  get: function () {
    return _OpenAPIParameter.default;
  }
});
Object.defineProperty(exports, "OpenAPIParameterMap", {
  enumerable: true,
  get: function () {
    return _OpenAPIParameterMap.default;
  }
});
Object.defineProperty(exports, "OpenAPIPath", {
  enumerable: true,
  get: function () {
    return _OpenAPIPath.default;
  }
});
Object.defineProperty(exports, "OpenAPIPaths", {
  enumerable: true,
  get: function () {
    return _OpenAPIPaths.default;
  }
});
Object.defineProperty(exports, "OpenAPIRequestBody", {
  enumerable: true,
  get: function () {
    return _OpenAPIRequestBody.default;
  }
});
Object.defineProperty(exports, "OpenAPIRequestBodyMap", {
  enumerable: true,
  get: function () {
    return _OpenAPIRequestBodyMap.default;
  }
});
Object.defineProperty(exports, "OpenAPIResponse", {
  enumerable: true,
  get: function () {
    return _OpenAPIResponse.default;
  }
});
Object.defineProperty(exports, "OpenAPIResponseMap", {
  enumerable: true,
  get: function () {
    return _OpenAPIResponseMap.default;
  }
});
Object.defineProperty(exports, "OpenAPIRoot", {
  enumerable: true,
  get: function () {
    return _OpenAPIRoot.default;
  }
});
Object.defineProperty(exports, "OpenAPISchema", {
  enumerable: true,
  get: function () {
    return _OpenAPISchema.default;
  }
});
Object.defineProperty(exports, "OpenAPISchemaMap", {
  enumerable: true,
  get: function () {
    return _OpenAPISchemaMap.default;
  }
});
Object.defineProperty(exports, "OpenAPISecurityRequirement", {
  enumerable: true,
  get: function () {
    return _OpenAPISecurityRequirement.default;
  }
});
Object.defineProperty(exports, "OpenAPISecuritySchema", {
  enumerable: true,
  get: function () {
    return _OpenAPISecuritySchema.default;
  }
});
Object.defineProperty(exports, "OpenAPISecuritySchemaMap", {
  enumerable: true,
  get: function () {
    return _OpenAPISecuritySchemaMap.default;
  }
});
Object.defineProperty(exports, "OpenAPISecuritySchemas", {
  enumerable: true,
  get: function () {
    return _OpenAPISecuritySchemas.default;
  }
});
Object.defineProperty(exports, "OpenAPIServer", {
  enumerable: true,
  get: function () {
    return _OpenAPIServer.default;
  }
});
Object.defineProperty(exports, "OpenAPIServerVariable", {
  enumerable: true,
  get: function () {
    return _OpenAPIServerVariable.default;
  }
});
Object.defineProperty(exports, "OpenAPIServerVariableMap", {
  enumerable: true,
  get: function () {
    return _OpenAPIServerVariableMap.default;
  }
});
Object.defineProperty(exports, "OpenAPITag", {
  enumerable: true,
  get: function () {
    return _OpenAPITag.default;
  }
});
Object.defineProperty(exports, "OpenAPIXML", {
  enumerable: true,
  get: function () {
    return _OpenAPIXML.default;
  }
});
Object.defineProperty(exports, "PasswordOpenAPIFlow", {
  enumerable: true,
  get: function () {
    return _PasswordOpenAPIFlow.default;
  }
});
Object.defineProperty(exports, "XRedocCodeSample", {
  enumerable: true,
  get: function () {
    return _XRedocCodeSample.default;
  }
});

var _AuthorizationCodeOpenAPIFlow = _interopRequireDefault(require("./AuthorizationCodeOpenAPIFlow"));

var _ClientCredentialsOpenAPIFlow = _interopRequireDefault(require("./ClientCredentialsOpenAPIFlow"));

var _ImplicitOpenAPIFlow = _interopRequireDefault(require("./ImplicitOpenAPIFlow"));

var _OpenAPICallback = _interopRequireDefault(require("./OpenAPICallback"));

var _OpenAPICallbackMap = _interopRequireDefault(require("./OpenAPICallbackMap"));

var _OpenAPIComponents = _interopRequireDefault(require("./OpenAPIComponents"));

var _OpenAPIContact = _interopRequireDefault(require("./OpenAPIContact"));

var _OpenAPIDiscriminator = _interopRequireDefault(require("./OpenAPIDiscriminator"));

var _OpenAPIEncoding = _interopRequireDefault(require("./OpenAPIEncoding"));

var _OpenAPIEncodingsMap = _interopRequireDefault(require("./OpenAPIEncodingsMap"));

var _OpenAPIExample = _interopRequireDefault(require("./OpenAPIExample"));

var _OpenAPIExampleMap = _interopRequireDefault(require("./OpenAPIExampleMap"));

var _OpenAPIExternalDocumentation = _interopRequireDefault(require("./OpenAPIExternalDocumentation"));

var _OpenAPIFlows = _interopRequireDefault(require("./OpenAPIFlows"));

var _OpenAPIHeader = _interopRequireDefault(require("./OpenAPIHeader"));

var _OpenAPIHeaderMap = _interopRequireDefault(require("./OpenAPIHeaderMap"));

var _OpenAPIInfo = _interopRequireDefault(require("./OpenAPIInfo"));

var _OpenAPILicense = _interopRequireDefault(require("./OpenAPILicense"));

var _OpenAPILink = _interopRequireDefault(require("./OpenAPILink"));

var _OpenAPILinkMap = _interopRequireDefault(require("./OpenAPILinkMap"));

var _OpenAPIMediaObject = _interopRequireDefault(require("./OpenAPIMediaObject"));

var _OpenAPIMediaTypeObject = _interopRequireDefault(require("./OpenAPIMediaTypeObject"));

var _OpenAPIOperation = _interopRequireDefault(require("./OpenAPIOperation"));

var _OpenAPIParameter = _interopRequireDefault(require("./OpenAPIParameter"));

var _OpenAPIParameterMap = _interopRequireDefault(require("./OpenAPIParameterMap"));

var _OpenAPIPath = _interopRequireDefault(require("./OpenAPIPath"));

var _OpenAPIPaths = _interopRequireDefault(require("./OpenAPIPaths"));

var _OpenAPIRequestBody = _interopRequireDefault(require("./OpenAPIRequestBody"));

var _OpenAPIRequestBodyMap = _interopRequireDefault(require("./OpenAPIRequestBodyMap"));

var _OpenAPIResponse = _interopRequireDefault(require("./OpenAPIResponse"));

var _OpenAPIResponseMap = _interopRequireDefault(require("./OpenAPIResponseMap"));

var _OpenAPIRoot = _interopRequireDefault(require("./OpenAPIRoot"));

var _OpenAPISchema = _interopRequireDefault(require("./OpenAPISchema"));

var _OpenAPISchemaMap = _interopRequireDefault(require("./OpenAPISchemaMap"));

var _OpenAPISecurityRequirement = _interopRequireDefault(require("./OpenAPISecurityRequirement"));

var _OpenAPISecuritySchema = _interopRequireDefault(require("./OpenAPISecuritySchema"));

var _OpenAPISecuritySchemaMap = _interopRequireDefault(require("./OpenAPISecuritySchemaMap"));

var _OpenAPISecuritySchemas = _interopRequireDefault(require("./OpenAPISecuritySchemas"));

var _OpenAPIServer = _interopRequireDefault(require("./OpenAPIServer"));

var _OpenAPIServerVariable = _interopRequireDefault(require("./OpenAPIServerVariable"));

var _OpenAPIServerVariableMap = _interopRequireDefault(require("./OpenAPIServerVariableMap"));

var _OpenAPITag = _interopRequireDefault(require("./OpenAPITag"));

var _OpenAPIXML = _interopRequireDefault(require("./OpenAPIXML"));

var _PasswordOpenAPIFlow = _interopRequireDefault(require("./PasswordOpenAPIFlow"));

var _XRedocCodeSample = _interopRequireDefault(require("../vendor/XRedocCodeSample"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
