// @redocly/openapi-cli/dist/types/OAS3/OpenAPIHeader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.OpenAPIHeader = void 0;

var _OpenAPIExampleMap = require("./OpenAPIExampleMap");

var _OpenAPIMediaTypeObject = require("./OpenAPIMediaTypeObject");

var _OpenAPISchema = _interopRequireDefault(require("./OpenAPISchema"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// eslint-disable-next-line import/no-cycle
const OpenAPIHeader = {
  name: 'OpenAPIHeader',
  isIdempotent: true,
  properties: {
    description: null,
    required: null,
    deprecated: null,
    allowEmptyValue: null,
    explode: null,
    allowReserved: null,
    example: null,
    schema: _OpenAPISchema.default,
    content: _OpenAPIMediaTypeObject.OpenAPIMediaTypeObject,
    examples: _OpenAPIExampleMap.OpenAPIExampleMap
  },
  resolvableScalars: ['description']
};
exports.OpenAPIHeader = OpenAPIHeader;
var _default = OpenAPIHeader;
exports.default = _default;
