// @redocly/openapi-cli/dist/types/OAS3/OpenAPIComponents.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _OpenAPISchemaMap = _interopRequireDefault(require("./OpenAPISchemaMap"));

var _OpenAPISecuritySchemaMap = _interopRequireDefault(require("./OpenAPISecuritySchemaMap"));

var _OpenAPIExampleMap = require("./OpenAPIExampleMap");

var _OpenAPIParameterMap = require("./OpenAPIParameterMap");

var _OpenAPIResponseMap = require("./OpenAPIResponseMap");

var _OpenAPIHeaderMap = require("./OpenAPIHeaderMap");

var _OpenAPILinkMap = require("./OpenAPILinkMap");

var _OpenAPICallbackMap = require("./OpenAPICallbackMap");

var _OpenAPIRequestBodyMap = require("./OpenAPIRequestBodyMap");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = {
  name: 'OpenAPIComponents',
  isIdempotent: true,
  properties: {
    schemas: _OpenAPISchemaMap.default,
    responses: _OpenAPIResponseMap.OpenAPIResponseMap,
    parameters: _OpenAPIParameterMap.OpenAPIParameterMap,
    examples: _OpenAPIExampleMap.OpenAPIExampleMap,
    requestBodies: _OpenAPIRequestBodyMap.OpenAPIRequestBodyMap,
    headers: _OpenAPIHeaderMap.OpenAPIHeaderMap,
    securitySchemes: _OpenAPISecuritySchemaMap.default,
    links: _OpenAPILinkMap.OpenAPILinkMap,
    callbacks: _OpenAPICallbackMap.OpenAPICallbackMap
  }
};
exports.default = _default;
