// @redocly/openapi-cli/dist/types/OAS3/OpenAPIOperation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _OpenAPIResponseMap = require("./OpenAPIResponseMap");

var _OpenAPIParameter = require("./OpenAPIParameter");

var _OpenAPIServer = _interopRequireDefault(require("./OpenAPIServer"));

var _OpenAPIExternalDocumentation = _interopRequireDefault(require("./OpenAPIExternalDocumentation"));

var _OpenAPICallbackMap = require("./OpenAPICallbackMap");

var _OpenAPIRequestBody = require("./OpenAPIRequestBody");

var _XRedocCodeSample = require("../vendor/XRedocCodeSample");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable import/no-cycle */
var _default = {
  name: 'OpenAPIOperation',
  isIdempotent: false,
  properties: {
    tags: null,
    summary: null,
    description: null,
    operationId: null,
    deprecated: null,
    security: null,
    externalDocs: _OpenAPIExternalDocumentation.default,
    parameters: _OpenAPIParameter.OpenAPIParameter,
    requestBody: _OpenAPIRequestBody.OpenAPIRequestBody,
    responses: _OpenAPIResponseMap.OpenAPIResponseMap,
    callbacks: _OpenAPICallbackMap.OpenAPICallbackMap,
    // TODO:
    // security() {},
    servers: _OpenAPIServer.default,
    'x-code-samples': _XRedocCodeSample.XRedocCodeSample
  },
  resolvableScalars: ['description']
};
exports.default = _default;
