// @redocly/openapi-cli/dist/types/OAS3/OpenAPIPath.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.OpenAPIPathItem = void 0;

var _OpenAPIServer = _interopRequireDefault(require("./OpenAPIServer"));

var _OpenAPIOperation = _interopRequireDefault(require("./OpenAPIOperation"));

var _OpenAPIParameter = require("./OpenAPIParameter");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable import/no-cycle */
const OpenAPIPathItem = {
  name: 'OpenAPIPath',
  isIdempotent: true,
  properties: {
    summary: null,
    description: null,
    parameters: _OpenAPIParameter.OpenAPIParameter,
    get: _OpenAPIOperation.default,
    put: _OpenAPIOperation.default,
    post: _OpenAPIOperation.default,
    delete: _OpenAPIOperation.default,
    options: _OpenAPIOperation.default,
    head: _OpenAPIOperation.default,
    patch: _OpenAPIOperation.default,
    trace: _OpenAPIOperation.default,
    servers: _OpenAPIServer.default
  },
  resolvableScalars: ['description']
};
exports.OpenAPIPathItem = OpenAPIPathItem;
var _default = OpenAPIPathItem;
exports.default = _default;
