// @redocly/openapi-cli/dist/types/OAS3/OpenAPIRoot.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.OpenAPIRoot = void 0;

var _OpenAPIInfo = require("./OpenAPIInfo");

var _OpenAPIPaths = require("./OpenAPIPaths");

var _OpenAPIComponents = _interopRequireDefault(require("./OpenAPIComponents"));

var _OpenAPIServer = _interopRequireDefault(require("./OpenAPIServer"));

var _OpenAPITag = _interopRequireDefault(require("./OpenAPITag"));

var _OpenAPIExternalDocumentation = _interopRequireDefault(require("./OpenAPIExternalDocumentation"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// import OpenAPISecurityRequirement from './OpenAPISecurityRequirement';
const OpenAPIRoot = {
  name: 'OpenAPIRoot',
  isIdempotent: true,
  properties: {
    openapi: null,
    security: null,
    info: _OpenAPIInfo.OpenAPIInfo,
    paths: _OpenAPIPaths.OpenAPIPaths,
    servers: _OpenAPIServer.default,
    components: _OpenAPIComponents.default,
    // security: OpenAPISecurityRequirement,
    tags: _OpenAPITag.default,
    externalDocs: _OpenAPIExternalDocumentation.default
  }
};
exports.OpenAPIRoot = OpenAPIRoot;
var _default = OpenAPIRoot;
exports.default = _default;
