// @redocly/openapi-cli/dist/types/OAS3/OpenAPIParameter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.OpenAPIParameter = void 0;

var _OpenAPISchema = _interopRequireDefault(require("./OpenAPISchema"));

var _OpenAPIMediaTypeObject = require("./OpenAPIMediaTypeObject");

var _OpenAPIExampleMap = require("./OpenAPIExampleMap");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const OpenAPIParameter = {
  name: 'OpenAPIParameter',
  isIdempotent: false,
  properties: {
    name: null,
    in: null,
    description: null,
    required: null,
    deprecated: null,
    allowEmptyValue: null,
    style: null,
    explode: null,
    allowReserved: null,
    example: null,
    schema: _OpenAPISchema.default,
    content: _OpenAPIMediaTypeObject.OpenAPIMediaTypeObject,
    examples: _OpenAPIExampleMap.OpenAPIExampleMap
  },
  resolvableScalars: ['description']
};
exports.OpenAPIParameter = OpenAPIParameter;
var _default = OpenAPIParameter;
exports.default = _default;
