// @redocly/openapi-cli/dist/types/OAS3/OpenAPIDiscriminator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.MAPPING_DATA_KEY = void 0;

var _traverse = _interopRequireDefault(require("../../traverse.js"));

var _OpenAPISchema = _interopRequireDefault(require("./OpenAPISchema"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable import/no-cycle */
const MAPPING_DATA_KEY = 'x-openapiCli_resolveMappingData';
exports.MAPPING_DATA_KEY = MAPPING_DATA_KEY;
var _default = {
  name: 'OpenAPIDiscriminator',
  isIdempotent: true,
  properties: {
    propertyName: null,
    mapping: null
  },
  customResolveFields: async (node, ctx, visited) => {
    if (node && node.mapping && typeof node.mapping === 'object') {
      ctx.path.push('mapping');

      for (const key of Object.keys(node.mapping)) {
        ctx.path.push(key);
        await (0, _traverse.default)({
          $ref: node.mapping[key],
          [MAPPING_DATA_KEY]: {
            // FIXME: too hacky
            mapping: node.mapping,
            key
          }
        }, _OpenAPISchema.default, ctx, visited);
        ctx.path.pop();
      }

      ctx.path.pop();
    }
  }
};
exports.default = _default;
