// @redocly/openapi-cli/dist/context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _path = _interopRequireDefault(require("path"));

var _loader = _interopRequireWildcard(require("./loader"));

var _utils = _interopRequireDefault(require("./visitors/utils"));

var _resolveDefinition = require("./resolveDefinition");

var _error = require("./error");

var _resolver = require("./resolver");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable no-param-reassign */
const validateFieldsRaw = (node, ctx, config, ruleName, validators) => {
  const result = [];
  const vals = Object.keys(validators);

  for (let i = 0; i < vals.length; i += 1) {
    if ((0, _utils.default)(config, vals[i])) {
      if (validators[vals[i]]) {
        ctx.path.push(vals[i]);
        const validate = validators[vals[i]].bind({
          rule: ruleName,
          config
        });
        const res = validate(node, ctx, config);

        if (res) {
          if (Array.isArray(res)) result.push(...res);else result.push(res);
        }

        ctx.path.pop();
      }
    }
  }

  return result;
};

const getRule = (ctx, ruleName) => {
  const result = ctx.allRules.filter(r => r.constructor.rule === ruleName);
  return result ? result[0] : null;
};

function createContext(node, sourceFile, filePath, config) {
  const [enabledRules, allRules] = (0, _loader.default)(config);
  config.headers = config.headers || [];
  return {
    openapiVersion: node.swagger ? 2 : 3,
    document: node,
    filePath: _path.default.resolve(filePath),
    path: [],
    cache: {},
    visited: [],
    result: [],
    registryDependencies: [],
    definitionStack: [],
    definitions: (0, _resolveDefinition.loadDefinitions)(config),
    pathStack: [],
    source: sourceFile,
    enableCodeframe: !!(config && (config.codeframes === 'on' || config.codeframes === true)),
    customRules: [...(0, _loader.loadRulesetExtension)(config, 'transformingVisitors'), ...enabledRules, ...(0, _loader.loadRulesetExtension)(config, 'rulesExtensions')],
    allRules,
    config,
    headers: config.headers,
    messageHelpers: _error.messageHelpers,
    validateFieldsRaw,
    getRule,
    resolveNode: _resolver.resolveNodeNoSideEffects,
    fileDependencies: new Set(),
    resolveCache: {}
  };
}

var _default = createContext;
exports.default = _default;
